testbench/database.py

```diff
@@ -256,3 +260,93 @@ def insert_rewrite(self, rewrite):
         del self.rewrites[token]
+
+    # ==== RETRY_TESTS ==== #
+
+    @classmethod
+    def __to_serializeable_retry_test(cls, retry_test):
+        return {
+            "id": retry_test["id"],
+            "instructions": {
+                key: list(value) for key, value in retry_test["instructions"].items()
+            },
+            "completed": retry_test["completed"],
+        }
+
+    def insert_supported_methods(self, methods):
+        self.supported_methods.extend(methods)
+
+    def get_retry_test(self, retry_test_id):
+        retry_test = self.retry_tests.get(retry_test_id)
+        if retry_test is None:
+            testbench.error.notfound("Retry Test %s" % retry_test_id, context=None)
+        return self.__to_serializeable_retry_test(retry_test)
+
+    def __validate_instructions(self, instructions):
+        expected_instructions_format = [
+            testbench.common.retry_return_error_code.pattern,
+            testbench.common.retry_return_error_connection.pattern,
+            testbench.common.retry_return_error_after_bytes.pattern,
+        ]
+        search_expression = re.compile(
+            "(%s)" % ")|(".join(expected_instructions_format)
+        )
+        expected_schema = Schema(
+            {
+                And(Use(str), lambda n: n in self.supported_methods): [
+                    And(Use(str), lambda n: re.match(search_expression, n))
+                ]
+            }
+        )
+        try:
+            expected_schema.validate(instructions)
+        except SchemaError as e:
+            testbench.error.invalid("Invalid format used: %s" % e, context=None)
+
+    def insert_retry_test(self, instructions):
+        self.__validate_instructions(instructions)
+        retry_test_id = uuid.uuid4().hex
+        self.retry_tests[retry_test_id] = {
+            "id": retry_test_id,
+            "instructions": {
+                key: collections.deque(value) for key, value in instructions.items()
+            },
+            "completed": False,
+        }
+        return self.__to_serializeable_retry_test(self.retry_tests[retry_test_id])
+
+    def has_instructions_retry_test(self, retry_test_id, method):
+        self.get_retry_test(retry_test_id)
+        if len(self.retry_tests[retry_test_id]["instructions"].get(method, [])) > 0:
+            return True
+        return False
+
+    def peek_next_instruction(self, retry_test_id, method):
+        self.get_retry_test(retry_test_id)
+        if self.retry_tests[retry_test_id]["instructions"] and self.retry_tests[
+            retry_test_id
+        ]["instructions"].get(method, None):
+            return self.retry_tests[retry_test_id]["instructions"][method][0]
+        else:
+            return None
+
+    def dequeue_next_instruction(self, retry_test_id, method):
+        self.get_retry_test(retry_test_id)
+        next_instruction = self.retry_tests[retry_test_id]["instructions"][
+            method
+        ].popleft()
```

Comment:
I guess this will return an indexError if the queue is empty, which might translate into a 5xx for the flask response? Do we need to catch this somewhere? Not sure exactly how this function is being used.

Reply:
> Not sure exactly how this function is being used.

I am not sure either. I am learning about the retry support as I move the classes and implement tests for them :-)

As the code stands today, they check if there are instructions just before using them:

https://github.com/googleapis/google-cloud-cpp/blob/a6329c16e474fe19f0d873d67eed408b26785cc5/google/cloud/storage/emulator/utils/common.py#L382

I might have chosen to return `None` as an indication of "no more instructions", but that sort of cleanup should happen after we move the code.

Reply:
Updated #27 to track this.

Reply:
Sounds good!